File: SQL/Amend/SQLFieldValues.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

Namespace SQL

    Public Class SQLFieldValues
        Implements IEnumerable

        Protected pobjFields As ArrayList = New ArrayList

        Public Sub New()

            MyBase.New()

        End Sub

        Public Overridable Function Add() As SQLFieldValue

            Return Add(String.Empty, Nothing)

        End Function

        Public Overridable Function Add(ByVal objFieldAndValue As SQLFieldValue) As SQLFieldValue

            Return Add(objFieldAndValue.Name, objFieldAndValue.Value)

        End Function

        Public Overridable Function Add( _
            ByVal strDestinationFieldName As String, _
            ByVal objEqualToValue As Object) As SQLFieldValue

            Dim objSQLFieldValue As New SQLFieldValue(strDestinationFieldName, objEqualToValue)

            pobjFields.Add(objSQLFieldValue)

            Return objSQLFieldValue

        End Function

        Public Overridable Sub Add(ByVal objFieldValues As SQLFieldValues)

            For Each objFieldValue As SQLFieldValue In objFieldValues
                Me.Add(objFieldValue)
            Next

        End Sub

        Default Public ReadOnly Property Item(ByVal strFieldName As String) As SQLFieldValue
            Get

                If Not Me.Exists(strFieldName) Then
                    Throw New ArgumentException("Field '" & strFieldName & "' does not exist.")
                End If

                Return DirectCast(pobjFields.Item(FieldNameIndex(strFieldName)), SQLFieldValue)

            End Get
        End Property

        Default Public ReadOnly Property Item(ByVal intIndex As Integer) As SQLFieldValue
            Get

                Return DirectCast(pobjFields.Item(intIndex), SQLFieldValue)

            End Get
        End Property

        Public ReadOnly Property Count() As Integer
            Get

                Return pobjFields.Count()

            End Get
        End Property

        Public Function Exists(ByVal strFieldName As String) As Boolean

            Return FieldNameIndex(strFieldName) >= 0

        End Function

        Public Sub Delete(ByRef objFieldValue As SQLFieldValue)

            If Not pobjFields.Contains(objFieldValue) Then
                Throw New IndexOutOfRangeException
            End If

            pobjFields.Remove(objFieldValue)
            objFieldValue = Nothing

        End Sub

        Friend Function FieldNameIndex(ByVal strFieldName As String) As Integer

            Dim objSQLFieldValue As SQLFieldValue

            strFieldName = strFieldName.Trim()

            For intIndex As Integer = 0 To Me.Count - 1
                objSQLFieldValue = DirectCast(pobjFields.Item(intIndex), SQLFieldValue)
                If String.Compare(strFieldName, objSQLFieldValue.Name, ignoreCase:=True) = 0 Then
                    Return intIndex
                End If
            Next

            Return -1

        End Function

        Private Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator

            Return pobjFields.GetEnumerator

        End Function

    End Class

End Namespace

File: SQL/Amend/SQLFieldValues.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;
using System.Collections.Generic;
using System.Linq;

namespace DatabaseObjects.SQL
{
	public class SQLFieldValues : IEnumerable<SQLFieldValue>
	{
		protected List<SQLFieldValue> pobjFields = new List<SQLFieldValue>();
			
		public SQLFieldValues()
		{
		}
			
		public virtual SQLFieldValue Add()
		{
			return Add(string.Empty, null);
		}
			
		public virtual SQLFieldValue Add(SQLFieldValue objFieldAndValue)
		{
			return Add(objFieldAndValue.Name, objFieldAndValue.Value);
		}
			
		public virtual SQLFieldValue Add(string strDestinationFieldName, object objEqualToValue)
		{
			SQLFieldValue objSQLFieldValue = new SQLFieldValue(strDestinationFieldName, objEqualToValue);
				
			pobjFields.Add(objSQLFieldValue);
				
			return objSQLFieldValue;
		}
			
		public virtual void Add(SQLFieldValues objFieldValues)
		{
			foreach (SQLFieldValue objFieldValue in objFieldValues)
				this.Add(objFieldValue);
		}
			
		public SQLFieldValue this[string strFieldName]
		{
			get
			{
				if (!this.Exists(strFieldName))
					throw new ArgumentException("Field '" + strFieldName + "' does not exist.");

				return pobjFields.First(field => Equals(field, strFieldName));
			}
		}
			
		public SQLFieldValue this[int intIndex]
		{
			get
			{
				return (SQLFieldValue)pobjFields[intIndex];
			}
		}
			
		public int Count
		{
			get
			{
				return pobjFields.Count;
			}
		}
			
		public bool Exists(string strFieldName)
		{
			return pobjFields.FirstOrDefault(field => Equals(field, strFieldName)) != null;
		}
			
		public void Delete(ref SQLFieldValue objFieldValue)
		{
			if (!pobjFields.Contains(objFieldValue))
				throw new IndexOutOfRangeException();
				
			pobjFields.Remove(objFieldValue);
			objFieldValue = null;
		}
			
		private bool Equals(SQLFieldValue fieldValue, string strFieldName)
		{
			return fieldValue.Name.Equals(strFieldName, StringComparison.InvariantCultureIgnoreCase);
		}
			
		IEnumerator<SQLFieldValue> IEnumerable<SQLFieldValue>.GetEnumerator()
		{
			return pobjFields.GetEnumerator();
		}

		IEnumerator IEnumerable.GetEnumerator()
		{
			return pobjFields.GetEnumerator();
		}
	}
}
